### ISSUE TITLE
Unable to connect to the Lexicon app

### ISSUE THREAD

**Issue by muninderkaur06** [lexiconhq/lexicon#4]
Hi, I have been trying to follow your docs to configure the prose instances for lexicon and discourse API but I'm unable to do so. I have tried several different combinations that I think were suitable for MOBILE_PROSE_HOST and MOBILE_PROSE_PORT. I even added the values http://localhost and 8999 respectively, but it keeps showing me the error 'Please connect to a network.'

Please help me fix this issue.

![image](https://user-images.githubusercontent.com/34957859/161455343-a3e374ee-1e23-4cf7-aca5-415280fe1a0b.png)

**Comment by Joshagos** [comment]
> Hi, I have been trying to follow your docs to configure the prose instances for lexicon and discourse API but I'm unable to do so. I have tried several different combinations that I think were suitable for MOBILE_PROSE_HOST and MOBILE_PROSE_PORT. I even added the values http://localhost and 8999 respectively, but it keeps showing me the error 'Please connect to a network.'
> 
> 
> 
> Please help me fix this issue.
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/34957859/161455343-a3e374ee-1e23-4cf7-aca5-415280fe1a0b.png)
> 
> 

I don't think this project is still being actively worked on.

**Comment by muninderkaur06** [comment]
Yeah, it's been months now. But they have instructed on how to configure, I did exactly the same still I am facing this issue.

**Comment by Joshagos** [comment]
> Yeah, it's been months now. But they have instructed on how to configure, I did exactly the same still I am facing this issue.

I wish they would explain why they've disappeared...

**Comment by akinseye21** [comment]
I have the same issue. After pointing to my discourse site, it keeps loading the default which is "meta.discourse.org" and sometimes tells me to connect to a network. Any help??

**Comment by RabanFabre** [comment]
> Hi, I have been trying to follow your docs to configure the prose instances for lexicon and discourse API but I'm unable to do so. I have tried several different combinations that I think were suitable for MOBILE_PROSE_HOST and MOBILE_PROSE_PORT. I even added the values http://localhost and 8999 respectively, but it keeps showing me the error 'Please connect to a network.'
> 
> Please help me fix this issue.

@muninderkaur06 in `api/.env`, try adding:

```
PROSE_APP_HOSTNAME=0.0.0.0
```

**Comment by RabanFabre** [comment]
> I have the same issue. After pointing to my discourse site, it keeps loading the default which is "meta.discourse.org" and sometimes tells me to connect to a network. Any help??

@akinseye21 could you share more about your setup?

- What are the specific contents you have in `api/.env` and `frontend/.env`?
- What are the specific commands you're running to start the API and the frontend?
- Are you running in a simulator on your computer? Or are you using the Expo Go app on your phone?

Here is an example setup that points at https://users.rust-lang.org:

#### api/.env
```
PROSE_DISCOURSE_HOST=https://users.rust-lang.org
PROSE_APP_HOSTNAME=0.0.0.0
PROSE_APP_PORT=8999
```

#### frontend/.env
```
MOBILE_PROSE_HOST=http://localhost
MOBILE_PROSE_PORT=8999
```

**Note**
Make sure you `cd frontend/` before calling `yarn start`.

----

#### api Output
```
$ yarn --cwd api start
 
Attempting to reach your Discourse instance at https://users.rust-lang.org...
Your Discourse instance was reachable and valid.

-- Prose GraphQL Discourse API --
listening at http://0.0.0.0:8999
forwarding Discourse requests to https://users.rust-lang.org
```

#### frontend
```
$ cd frontend
$ yarn start
<Expo output here, scan QR code>
```

![181471176-ed128780-ed6a-4e8d-9a6d-4cd8bcf9434f](https://user-images.githubusercontent.com/110163380/181471527-ebe8de57-cc6a-4328-a05e-ee24b6e1d1fd.png)

**Comment by Kinostrome** [comment]
> > I have the same issue. After pointing to my discourse site, it keeps loading the default which is "meta.discourse.org" and sometimes tells me to connect to a network. Any help??
> 
> @akinseye21 could you share more about your setup?
> 
>     * What are the specific contents you have in `api/.env` and `frontend/.env`?
> 
>     * What are the specific commands you're running to start the API and the frontend?
> 
>     * Are you running in a simulator on your computer? Or are you using the Expo Go app on your phone?


Thanks for posting this, this is a good way to help others debug and is very appreciated.

**Comment by Kinostrome** [comment]
Coming up next:

- An upgrade! to Expo 45, as well as most other dependencies 🎉
- A fix for #1 
- An update on everything to help manage expectations going forward

**Comment by Kinostrome** [comment]
On this particular issue, since it seems to have caused some confusion, I have also [updated the defaults](https://github.com/lexiconhq/lexicon/blob/master/api/src/constants/api.ts#L12) in Prose.

This will cause Prose to listen on `0.0.0.0` by default. Meaning—for the scenario where you have the app running on Expo Go on your phone, and Prose running locally on your machine, the app should be able to locate Prose and connect to it without needing to specify `PROSE_APP_HOSTNAME=0.0.0.0`.

In this scenario, it is of course required that your laptop and mobile device are on the same network and can reach each other.

**Comment by muninderkaur06** [comment]
Hi people, 
My lexicon app is working perfectly now. 

What I did was I found out that IP address and giving full permissions to nodeJs are the most important thing here. 

I figured out, whenever the system IP address changes the app doesn't work. 
Like, whenever I am setting up lexicon, at that point whatever my IP address is it somehow binds with the app. In future if the address changes, app doesn't work.

To fix this, I have to a statically fix my IP address to the one my system is assigned during lexicon setup procedure. (For IPV4 in adapter settings.)

**Comment by muninderkaur06** [comment]
Let me mention,

For frontend folder in .env file:

MOBILE_PROSE_HOST=http://192.168.1.3 (Your system IPV4 address)
MOBILE_PROSE_PORT=80

For api folder in .env file:

PROSE_DISCOURSE_HOST=(meta discourse website)
MOBILE_PROSE_HOST=http://0.0.0.0
PROSE_APP_HOSTNAME=0.0.0.0

These values worked for me.

**Comment by Kinostrome** [comment]
> Hi people, My lexicon app is working perfectly now.
> 
> What I did was I found out that IP address and giving full permissions to nodeJs are the most important thing here.
> 
> I figured out, whenever the system IP address changes the app doesn't work. Like, whenever I am setting up lexicon, at that point whatever my IP address is it somehow binds with the app. In future if the address changes, app doesn't work.
> 
> To fix this, I have to a statically fix my IP address to the one my system is assigned during lexicon setup procedure. (For IPV4 in adapter settings.)


Hello,

Glad you were able to get it working.

We do cover this part about finding your local IP address the [Development Setup](https://docs.lexicon.is/setup#scenario-3-run-prose-locally--access-from-your-mobile-device) section of the documentation.

However, if you're just running Expo on your mobile device, and your development machine is on the same network, much of the config you posted shouldn't be necessary.

By the way, `MOBILE_PROSE_HOST` doesn't belong in `api/.env`. It is only a valid value in the frontend, so you should remove that from your `api/.env` file.

Importantly, if you set `frontend/.env` to have `MOBILE_PROSE_HOST=http://localhost` (or `http://127.0.0.1`), the frontend React Native app will automatically attempt to discover your development host (`debuggerHost`) through Expo. 

You can read more about how we determine `debuggerHost` in [this section](https://docs.lexicon.is/setup#brief-note-about-the-android-simulator) of the documentation.

i.e. I was just able to get everything up and running with this only this config:

`frontend/.env`
```
MOBILE_PROSE_HOST=http://127.0.0.1
```

`api/.env`
```
PROSE_DISCOURSE_HOST=https://springboot.io
```

#### Shell
```
$ git clone git@github.com:lexiconhq/lexicon.git lexicon-test
$ cd lexicon-test
$ yarn
$ yarn generate # Don't forget to run yarn generate when first cloning!
$ echo "PROSE_DISCOURSE_HOST=https://users.rust-lang.org" > api/.env
$ echo "MOBILE_PROSE_HOST=http://localhost" > frontend/.env
$ yarn --cwd api start & # API launched as a background process, or just do it in a separate shell

Attempting to reach your Discourse instance at https://users.rust-lang.org...
Your Discourse instance was reachable and valid.

-- Prose GraphQL Discourse API --
listening at http://0.0.0.0:80
forwarding Discourse requests to https://users.rust-lang.org

$ cd frontend
$ yarn start
$ expo start
Starting project at lexicon-test/frontend
Starting Metro Bundler
...

› Metro waiting on exp://192.168.0.53:19000
› Scan the QR code above with Expo Go (Android) or the Camera app (iOS)

...

Logs for your project will appear below. Press Ctrl+C to exit.
Started Metro Bundler

```

**Comment by Kinostrome** [comment]
For anyone interested, you can see how we're determining the local IP address of your development machine from within the React Native [here: src/frontend/constants/app.ts](https://github.com/lexiconhq/lexicon/blob/master/frontend/src/constants/app.ts#L30-L46)

```ts
  // ...

  // Only if we're not in production, and we're attempting to connect our local machine
  // (localhost or 127.0.0.1), extract the debuggerHost if possible and use that instead.
  // This allows the Android Emulator, as well as your own device on the same network, to
  // access your local machine to connect to a locally running Prose instance.

  const { debuggerHost } = manifest ?? {};
  if (!debuggerHost) {
    return defaultValue;
  }

  // Example `debuggerHost`: 192.168.0.53:19000
  const machineHost = debuggerHost.split(':')[0];
  if (!machineHost) {
    return defaultValue;
  }

  return `http://${machineHost}${delimitedProsePort}`;
}

// ...
```

**Comment by muninderkaur06** [comment]
@Kinostrome It's how my app is been working with these values in .env file.

Please do mention somewhere that the IP address needs to be same. 

If today I'm doing the lexicon setup and my IP address is suppose http://192.168.1.3, it may work today. 
Tomorrow if my system address changes to http://192.168.1.5, even after changing .env file in frontend App won't run. I will have to do the setup again.

**Comment by Kinostrome** [comment]
> @Kinostrome It's how my app is been working with these values in .env file.
> 
> Please do mention somewhere that the IP address needs to be same.
> 
> If today I'm doing the lexicon setup and my IP address is suppose http://192.168.1.3, it may work today. Tomorrow if my system address changes to http://192.168.1.5, even after changing .env file in frontend App won't run. I will have to do the setup again.

Yes, please read what I wrote again. We have already set it up so you don't have to deal with this. This is exactly the reason that I said:

>Importantly, if you set frontend/.env to have MOBILE_PROSE_HOST=http://localhost (or http://127.0.0.1), the frontend React Native app **will _automatically_ attempt to discover your development host** (debuggerHost) through Expo.

You **do not** need to manually set your IP address if you set `MOBILE_PROSE_HOST=http://localhost`.

**Comment by akinseye21** [comment]
> > I have the same issue. After pointing to my discourse site, it keeps loading the default which is "meta.discourse.org" and sometimes tells me to connect to a network. Any help??
> 
> @akinseye21 could you share more about your setup?
> 
> * What are the specific contents you have in `api/.env` and `frontend/.env`?
> * What are the specific commands you're running to start the API and the frontend?
> * Are you running in a simulator on your computer? Or are you using the Expo Go app on your phone?
> 
> Here is an example setup that points at https://users.rust-lang.org:
> 
> #### api/.env
> ```
> PROSE_DISCOURSE_HOST=https://users.rust-lang.org
> PROSE_APP_HOSTNAME=0.0.0.0
> PROSE_APP_PORT=8999
> ```
> 
> #### frontend/.env
> ```
> MOBILE_PROSE_HOST=http://localhost
> MOBILE_PROSE_PORT=8999
> ```
> 
> **Note** Make sure you `cd frontend/` before calling `yarn start`.
> 
> #### api Output
> ```
> $ yarn --cwd api start
>  
> Attempting to reach your Discourse instance at https://users.rust-lang.org...
> Your Discourse instance was reachable and valid.
> 
> -- Prose GraphQL Discourse API --
> listening at http://0.0.0.0:8999
> forwarding Discourse requests to https://users.rust-lang.org
> ```
> 
> #### frontend
> ```
> $ cd frontend
> $ yarn start
> <Expo output here, scan QR code>
> ```
> 
> ![181471176-ed128780-ed6a-4e8d-9a6d-4cd8bcf9434f](https://user-images.githubusercontent.com/110163380/181471527-ebe8de57-cc6a-4328-a05e-ee24b6e1d1fd.png)

Thank you, I will try this and revert

**Comment by Kinostrome** [comment]
> Thank you, I will try this and revert

Great, let us know how it goes.

Some tips:

- Make sure you `git pull` the latest changes down. We just pushed an update yesterday. Rerun `yarn install` and `yarn generate` from the project root just to be safe as well.

- If you change the values in `frontend/.env`, such as the port or host of the Prose server, and nothing is updating, try running `expo r -c` (after running `cd frontend/`) to clear the cache and see if that fixes it.

**Comment by akinseye21** [comment]
> > Hi people, My lexicon app is working perfectly now.
> > What I did was I found out that IP address and giving full permissions to nodeJs are the most important thing here.
> > I figured out, whenever the system IP address changes the app doesn't work. Like, whenever I am setting up lexicon, at that point whatever my IP address is it somehow binds with the app. In future if the address changes, app doesn't work.
> > To fix this, I have to a statically fix my IP address to the one my system is assigned during lexicon setup procedure. (For IPV4 in adapter settings.)
> 
> Hello,
> 
> Glad you were able to get it working.
> 
> We do cover this part about finding your local IP address the [Development Setup](https://docs.lexicon.is/setup#scenario-3-run-prose-locally--access-from-your-mobile-device) section of the documentation.
> 
> However, if you're just running Expo on your mobile device, and your development machine is on the same network, much of the config you posted shouldn't be necessary.
> 
> By the way, `MOBILE_PROSE_HOST` doesn't belong in `api/.env`. It is only a valid value in the frontend, so you should remove that from your `api/.env` file.
> 
> Importantly, if you set `frontend/.env` to have `MOBILE_PROSE_HOST=http://localhost` (or `http://127.0.0.1`), the frontend React Native app will automatically attempt to discover your development host (`debuggerHost`) through Expo.
> 
> You can read more about how we determine `debuggerHost` in [this section](https://docs.lexicon.is/setup#brief-note-about-the-android-simulator) of the documentation.
> 
> i.e. I was just able to get everything up and running with this only this config:
> 
> `frontend/.env`
> 
> ```
> MOBILE_PROSE_HOST=http://127.0.0.1
> ```
> 
> `api/.env`
> 
> ```
> PROSE_DISCOURSE_HOST=https://springboot.io
> ```
> 
> #### Shell
> ```shell
> $ git clone git@github.com:lexiconhq/lexicon.git lexicon-test
> $ cd lexicon-test
> $ yarn
> $ yarn generate # Don't forget to run yarn generate when first cloning!
> $ echo "PROSE_DISCOURSE_HOST=https://users.rust-lang.org" > api/.env
> $ echo "MOBILE_PROSE_HOST=http://localhost" > frontend/.env
> $ yarn --cwd api start & # API launched as a background process, or just do it in a separate shell
> 
> Attempting to reach your Discourse instance at https://users.rust-lang.org...
> Your Discourse instance was reachable and valid.
> 
> -- Prose GraphQL Discourse API --
> listening at http://0.0.0.0:80
> forwarding Discourse requests to https://users.rust-lang.org
> 
> $ cd frontend
> $ yarn start
> Starting project at lexicon-test/frontend
> Starting Metro Bundler
> ...
> 
> › Metro waiting on exp://192.168.0.53:19000
> › Scan the QR code above with Expo Go (Android) or the Camera app (iOS)
> 
> ...
> 
> Logs for your project will appear below. Press Ctrl+C to exit.
> Started Metro Bundler
> ```

Hello guys!! this worked well for me. Thank you. One last thing, I am having issues building the apk file of the app. Any help?

**Comment by Kinostrome** [comment]
> Hello guys!! this worked well for me. Thank you. One last thing, I am having issues building the apk file of the app. Any help?

Really glad to hear that.

Did you run the command outlined in [Publishing to the Play Store](https://docs.lexicon.is/play-store#build-your-app-for-android)? After setting up the required Expo configuration.

```
$ expo build:android
```

If so, what kind of error messages are you seeing?

**Comment by akinseye21** [comment]
> > Hello guys!! this worked well for me. Thank you. One last thing, I am having issues building the apk file of the app. Any help?
> 
> Really glad to hear that.
> 
> Did you run the command outlined in [Publishing to the Play Store](https://docs.lexicon.is/play-store#build-your-app-for-android)? After setting up the required Expo configuration.
> 
> ```
> $ expo build:android
> ```
> 
> If so, what kind of error messages are you seeing?

Thank you for the help, it built successfully. I was initiating the command from the root folder rather than the frontend/

**Comment by akinseye21** [comment]
> > Hello guys!! this worked well for me. Thank you. One last thing, I am having issues building the apk file of the app. Any help?
> 
> Really glad to hear that.
> 
> Did you run the command outlined in [Publishing to the Play Store](https://docs.lexicon.is/play-store#build-your-app-for-android)? After setting up the required Expo configuration.
> 
> ```
> $ expo build:android
> ```
> 
> If so, what kind of error messages are you seeing?

My build not working appropriately. Does my mobile device have to be on the same network with my computer for it to function?

**Comment by Kinostrome** [comment]
> My build not working appropriately. Does my mobile device have to be on the same network with my computer for it to function?

### Please provide proper details

Unfortunately that is not enough detail for us to help you. "not working appropriately" doesn't tell us anything about what could be going wrong.

What is happening? What are you expecting to happen? Post console output, screenshots, videos, etc.

You will need to describe **how specifically** the build is not working appropriately. Otherwise there's not much more to be said here.

**Comment by akinseye21** [comment]
> > My build not working appropriately. Does my mobile device have to be on the same network with my computer for it to function?
> 
> ### Please provide proper details
> Unfortunately that is not enough detail for us to help you. "not working appropriately" doesn't tell us anything about what could be going wrong.
> 
> What is happening? What are you expecting to happen? Post console output, screenshots, videos, etc.
> 
> You will need to describe **how specifically** the build is not working appropriately. Otherwise there's not much more to be said here.

The app shows an error message  "Please connect to a network" when I try to login

**Comment by Kinostrome** [comment]
> The app shows an error message "Please connect to a network" when I try to login

Please try to provide more detail with these issue reports.. this doesn't tell us the important details like:
- If you're running the app on your device or on the simulator
- If on your device, whether it's an actual build or running through Expo

If you're trying to publish your app to the App Store / Play Store, then you need to think about how your users are going to access your Prose server—meaning you'll need to deploy it somewhere. After all, they wouldn't download the app to have it connect to the server running on your local computer, would they?

Supposing you wanted to deploy it to:

```
https://prose.akinseye21.com
```

Then your `api/.env` would need to have the value:
```
PROSE_DISCOURSE_HOST=https://your.discourse.site
```

And your `frontend/.env` would need to have the value:
```
MOBILE_PROSE_HOST=https://prose.akinseye21.com
```

With those values set, then you would build your app.

To get a proper understanding, you should read through the documentation about how to deploy the Prose GraphQL API here: https://docs.lexicon.is/dedicated

**Comment by Kinostrome** [comment]
Closing this issue. If anyone has further problems setting things up, please open a new issue.
